ShopperInsights/src/main/java/com/braintreepayments/api/ShopperInsightsClient.kt

```diff
@@ -16,18 +17,43 @@ import com.braintreepayments.api.ShopperInsightsAnalytics.VENMO_SELECTED
  */
 class ShopperInsightsClient @VisibleForTesting internal constructor(
     private val paymentReadyAPI: PaymentReadyApi,
-    private val braintreeClient: BraintreeClient
+    private val braintreeClient: BraintreeClient,
+    private val deviceInspector: DeviceInspector
 ) {
+    constructor(braintreeClient: BraintreeClient) : this(
+        PaymentReadyApi(),
+        braintreeClient,
+        DeviceInspector()
+    )
+
     /**
      * Retrieves recommended payment methods based on the provided shopper insights request.
      *
+     * @param context Android context
      * @param request The [ShopperInsightsRequest] containing information about the shopper.
      * @return A [ShopperInsightsResult] object indicating the recommended payment methods.
      */
     fun getRecommendedPaymentMethods(
+        context: Context,
         request: ShopperInsightsRequest,
         callback: ShopperInsightsCallback
     ) {
+        val applicationContext = context.applicationContext
+        val isVenmoAppInstalled = deviceInspector.isVenmoInstalled(applicationContext)
+        val isPayPalAppInstalled = deviceInspector.isPayPalInstalled(applicationContext)
+
+        if (isVenmoAppInstalled && isPayPalAppInstalled) {
+            callback.onResult(
+                ShopperInsightsResult.Success(
+                    ShopperInsightsInfo(
+                        isPayPalRecommended = true,
+                        isVenmoRecommended = true
+                    )
+                )
+            )
+            return
+        }
+
         if (request.email == null && request.phone == null) {
```

Comment:
Good question.
My line of thought was that we had sufficient information to give the merchants a result they need to display the buttons, so we'd return the result.

The focus was on the final result rather than the API consistency.
I don't know if we need to prefer one way or the other. Do we need to consult with product for a better understanding of how this needs to work?

Reply:
Nah - I think we can make this call ourselves. Whatever we think delivers the best merchant dev experience.

Reply:
what does iOS do in this case?

Reply:
Changed logic to do the validation first.

Reply:
On iOS, we require the email or phone in the init of the request obj, so we don't have to do the validation in the same was as Android